homeassistant/components/otbr/__init__.py

```diff
@@ -0,0 +1,222 @@
+"""The Open Thread Border Router integration."""
+from __future__ import annotations
+
+import dataclasses
+from http import HTTPStatus
+from json import JSONDecodeError
+
+import aiohttp
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN
+from .models import OperationalDataSet, ThreadState
+
+
+@dataclasses.dataclass
+class OTBRData:
+    """Container for OTBR data."""
+
+    host: str
+    port: str
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up an Open Thread Border Router config entry."""
+
+    hass.data[DOMAIN] = OTBRData(entry.data["host"], entry.data["port"])
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    hass.data.pop(DOMAIN)
+    return True
+
+
+class ThreadNetworkActiveError(HomeAssistantError):
+    """Raised on attempts to modify the active dataset when thread network is active."""
+
+
+class NoDatasetError(HomeAssistantError):
+    """Raised on attempts to update a dataset which does not exist."""
+
+
+async def _async_get_thread_rest_service_url(hass) -> str:
+    """Return Thread REST API URL."""
+    otbr_data: OTBRData | None = hass.data.get(DOMAIN)
+    if not otbr_data:
+        raise HomeAssistantError
+
+    return f"http://{otbr_data.host}:{otbr_data.port}"
+
+
+def _raise_for_status(response: aiohttp.ClientResponse) -> None:
+    """Raise if status >= 400."""
+    try:
+        response.raise_for_status()
+    except aiohttp.ClientResponseError as exc:
+        raise HomeAssistantError from exc
+
+
+async def async_get_thread_state(hass: HomeAssistant) -> ThreadState:
+    """Get current Thread state."""
+
+    response = await async_get_clientsession(hass).get(
```

Comment:
Can we make a helper class for the API that has a request wrapper to handle the raise for status?

https://github.com/esphome/dashboard-api/blob/main/esphome_dashboard_api/__init__.py#L42-L54